File: docs/examples/javascript/games/get-game-state.js
/**
 * Get Game State
 *
 * Retrieves the current state of a game including rounds, progress, and settings.
 *
 * Usage in browser console:
 * 1. Navigate to geoguessr.com (must be logged in)
 * 2. Paste this code into the console
 * 3. Call: getGameState('GAME_TOKEN')
 */

async function getGameState(gameToken) {
    try {
        const url = `https://www.geoguessr.com/api/v3/games/${gameToken}?client=web`;

        const response = await fetch(url, {
            credentials: 'include'
        });

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${response.statusText}`);
        }

        const game = await response.json();

        // Display game information
        console.log(`\n🎮 Game State\n`);
        console.log(`Token: ${game.token}`);
        console.log(`Mode: ${game.mode} (${game.type})`);
        console.log(`State: ${game.state}`);
        console.log(`Map: ${game.mapName}`);

        console.log(`\n⚙️  Settings:`);
        console.log(`Rounds: ${game.roundCount}`);
        console.log(`Time Limit: ${game.timeLimit === 0 ? 'None' : game.timeLimit + 's'}`);
        console.log(`Movement: ${game.forbidMoving ? '❌ Disabled' : '✅ Enabled'}`);
        console.log(`Panning: ${game.forbidRotating ? '❌ Disabled' : '✅ Enabled'}`);
        console.log(`Zooming: ${game.forbidZooming ? '❌ Disabled' : '✅ Enabled'}`);

        console.log(`\n📊 Progress:`);
        console.log(`Current Round: ${game.round} of ${game.roundCount}`);

        if (game.player) {
            const player = game.player;
            console.log(`\n👤 Player Stats:`);
            console.log(`Score: ${parseInt(player.totalScore.amount).toLocaleString()} points`);
            console.log(`Distance: ${player.totalDistanceInMeters.toLocaleString()}m`);
            console.log(`Time: ${player.totalTime}s`);
            console.log(`Guesses: ${player.guesses.length}`);

            if (game.mode === 'streak') {
                console.log(`Streak: ${player.totalStreak}`);
            }
        }

        // Show rounds
        if (game.rounds && game.rounds.length > 0) {
            console.log(`\n📍 Rounds:`);
            game.rounds.forEach((round, index) => {
                const status = game.player?.guesses[index] ? '✅' : '⏳';
                console.log(`${status} Round ${index + 1}: ${round.lat.toFixed(4)}, ${round.lng.toFixed(4)}`);
            });
        }

        // Show guesses if any
        if (game.player?.guesses && game.player.guesses.length > 0) {
            console.log(`\n🎯 Your Guesses:`);
            game.player.guesses.forEach((guess, index) => {
                console.log(`Round ${index + 1}: ${guess.roundScore.amount} points (${guess.distance.meters.amount}m away)`);
            });
        }

        return game;

    } catch (error) {
        console.error('❌ Error fetching game state:', error.message);
        return null;
    }
}

// Example usage:
// getGameState('XB9wycuZvDrhl7cB');

// Get game token from current URL:
// const currentToken = window.location.pathname.split('/').pop();
// getGameState(currentToken);

// Monitor game progress (poll every 5 seconds):
// async function monitorGame(token) {
//     while (true) {
//         const game = await getGameState(token);
//         if (!game || game.state === 'finished') break;
//         await new Promise(r => setTimeout(r, 5000));
//     }
// }

File: examples/python/games/get_game_state.py
"""
Get Game State

Retrieves the current state of a game including rounds, progress, and settings.

Usage:
    python get_game_state.py

Requirements:
    pip install requests
"""

import requests
import os
from datetime import datetime


def get_game_state(game_token, cookie):
    """
    Fetch and display game state.

    Args:
        game_token (str): Game token (16-character string)
        cookie (str): Your _ncfa cookie value

    Returns:
        dict: Game state object
    """
    try:
        url = f'https://www.geoguessr.com/api/v3/games/{game_token}?client=web'

        response = requests.get(url, cookies={'_ncfa': cookie})
        response.raise_for_status()

        game = response.json()

        # Display game information
        print(f"\n🎮 Game State\n")
        print(f"Token: {game['token']}")
        print(f"Mode: {game['mode']} ({game['type']})")
        print(f"State: {game['state']}")
        print(f"Map: {game['mapName']}")

        print(f"\n⚙️  Settings:")
        print(f"Rounds: {game['roundCount']}")
        time_limit = 'None' if game['timeLimit'] == 0 else f"{game['timeLimit']}s"
        print(f"Time Limit: {time_limit}")
        print(f"Movement: {'❌ Disabled' if game['forbidMoving'] else '✅ Enabled'}")
        print(f"Panning: {'❌ Disabled' if game['forbidRotating'] else '✅ Enabled'}")
        print(f"Zooming: {'❌ Disabled' if game['forbidZooming'] else '✅ Enabled'}")

        print(f"\n📊 Progress:")
        print(f"Current Round: {game['round']} of {game['roundCount']}")

        if 'player' in game and game['player']:
            player = game['player']
            print(f"\n👤 Player Stats:")
            print(f"Score: {int(player['totalScore']['amount']):,} points")
            print(f"Distance: {player['totalDistanceInMeters']:,}m")
            print(f"Time: {player['totalTime']}s")
            print(f"Guesses: {len(player['guesses'])}")

            if game['mode'] == 'streak':
                print(f"Streak: {player['totalStreak']}")

        # Show rounds
        if 'rounds' in game and game['rounds']:
            print(f"\n📍 Rounds:")
            for index, round_data in enumerate(game['rounds']):
                status = '✅' if game.get('player', {}).get('guesses', []) and len(game['player']['guesses']) > index else '⏳'
                print(f"{status} Round {index + 1}: {round_data['lat']:.4f}, {round_data['lng']:.4f}")

        # Show guesses if any
        if game.get('player', {}).get('guesses'):
            print(f"\n🎯 Your Guesses:")
            for index, guess in enumerate(game['player']['guesses']):
                score = int(guess['roundScore']['amount'])
                distance = int(guess['distance']['meters']['amount'])
                print(f"Round {index + 1}: {score:,} points ({distance:,}m away)")

        return game

    except requests.exceptions.HTTPError as e:
        print(f"❌ HTTP Error: {e}")
        if e.response.status_code == 401:
            print("Note: Your cookie may be invalid or expired")
        elif e.response.status_code == 404:
            print("Note: Game not found or you don't have access")
        return None
    except requests.exceptions.RequestException as e:
        print(f"❌ Request Error: {e}")
        return None


def main():
    # Get credentials from environment variables
    cookie = os.getenv('GEOGUESSR_COOKIE')
    if not cookie:
        print("❌ Error: GEOGUESSR_COOKIE environment variable not set")
        print("Set it with: export GEOGUESSR_COOKIE='your_cookie_value'")
        return

    # Get game token
    game_token = input("Enter game token: ").strip()

    if not game_token:
        print("❌ Game token cannot be empty")
        return

    # Fetch game state
    game = get_game_state(game_token, cookie)

    if game:
        print("\n✅ Game state retrieved successfully!")


if __name__ == '__main__':
    main()
